src/lib/operators/table_scan/abstract_dereferenced_column_table_scan_impl.cpp

```diff
@@ -19,11 +19,11 @@ AbstractDereferencedColumnTableScanImpl::AbstractDereferencedColumnTableScanImpl
     const PredicateCondition init_predicate_condition)
     : predicate_condition(init_predicate_condition), _in_table(in_table), _column_id(column_id) {}
 
-std::shared_ptr<PosList> AbstractDereferencedColumnTableScanImpl::scan_chunk(const ChunkID chunk_id) const {
+std::shared_ptr<RowIDPosList> AbstractDereferencedColumnTableScanImpl::scan_chunk(const ChunkID chunk_id) const {
   const auto chunk = _in_table->get_chunk(chunk_id);
   const auto& segment = chunk->get_segment(_column_id);
 
-  auto matches = std::make_shared<PosList>();
+  auto matches = std::make_shared<RowIDPosList>();
```

Comment:
I like the idea. All _scan_non_reference_segments seem to end up in `_scan_with_iterators` or `SegmentSearch.scan_sorted_segment`. It's a bit tricky though as for some paths, I think all the current interfaces have to continue existing (allow passing a pre-filled RowIDPosList to the _scan_ methods). We could use templating to get this without much duplication, but we would have to move a lot of methods to the .hpp files then, probably increasing compilation and especially recompilation times.

This is also applicable in some operators. For example in the validate operator, there is a path where we know that all `RowID`s have the same `chunk_id` but we're creating a `RowIDPosList`.

Reply:
(See https://github.com/hyrise/hyrise/pull/2026#discussion_r400113202 - @XPerianer had almost zero perfomance impact with the SingleChunkPosList, which I don't really understand)

Reply:
Could you please commit this (and revert it) so that I can have a look at the diff?

Reply:
Perhaps I should have been clearer: he wasn't benchmarking what we are discussing here, he was just testing the performance impact on the index scan, so he was on 08220ad2a

Reply:
Actually, on 7aaeb99, with the modification that the `l_shipdate` index is created. I am also pretty sure that the indexScan was executed.

Using a higher scale factor (3) was actually slowing down the version with index scan more, which is why I think we have another bottleneck here that is killing the performance gain.

Reply:
I remember that we took some measurements on where the IndexScan spends its time. I believe it was something like 99% in the loop that writes the PosList. If we only change the loop to add ChunkOffsets instead of RowIDs and do not fundamentally change that behavior, I would not expect any performance improvements.

Reply:
> I would not expect any performance improvements.

Why? If we reduce the amount of data that has to be written by factor two, I'd also expect this memory-write-loop to speed up by factor two since I would expect this to be bottlenecked by memory write speed.
(I just verified that on vm-nordes, the RowIDPosList uses 8 bytes per element and the SingleChunkPosList uses 4 bytes per element, so there's no padding going on that could slow us down)

Reply:
But during the course of your project we have never seen any performance improvements close to that, have we? In addition, it's not only the writing that consumes time.

In the past we have even seen operations on 32bit data types to be slower than on 64bit data types. Reasons could be manifold, e.g., compiler or hardware optimizations. A small experiment on my machine does not show any significant differences between `uint32_t` and `uint64_t` vectors. By verified you mean the width that is used per element or the speed you can achieve?

I don't even know if a 64bit system would not actually handle a write to a 32bit (or smaller) value as a 64bit operation. Probably, @mrks knows much more about all of that.

Reply:
No, the system would still have to write 32 bits (otherwise, you would overwrite other data). However, for steps that are not memory-bound (and single-threaded operations often are not entirely memory-bound), a lot of effects come into play. There is a good chance that the benchmarks tell a different story when you use multiple cores and increase the amount of data. In any case, I have given up making any claims about these things. I'd form a hypothesis, run benchmarks, and try to understand the results.

Reply:
> A small experiment on my machine does not show any significant differences between `uint32_t` and `uint64_t` vectors. By verified you mean the width that is used per element or the speed you can achieve?

@Bensk1 Can you put what you tested in a gist / pastebin / whatever?

I tried this: https://gist.github.com/He3lixxx/bb21cb1f21b48d215e9ecdc123088568
and it gave me exactly what I would expect:
```
g++ -O3 on vm-nordes:
RowIDPosList took 5031ms.
SingleChunkPosList took 1924ms.

clang++ -O3 on vm-nordes:
RowIDPosList took 8085ms. // even 4x slower!
SingleChunkPosList took 1964ms.
```

(For me, just filling a vector with some uint32_t is about twice as fast as filling it with some uint64_t - see http://quick-bench.com/nUHLbcdLoGHVOg6wPZPHqAObJws. Is it possible that some of your code was simply optimized away?)

The "verified" was about the SingleChunkPosList using 4 bytes per element in the offsets vector (no padding is added - if this was padded to 8 bytes, it would make sense to me that we don't see better performance)

With the tests I've done so far, I'd expect that if filling the RowIDPosList really is the slow part, making it a SingleChunkPosList should double the performance of this part of the code.
On the index scan, if replacing it does not yield performance improvements, I'd say either traversing the index is slow or we get a slow down when we later use that PosList.

Reply:
I also just filled a vector and wanted to use quick-bench for that which seems to have some kind of server problem. As the test loop took a couple of hundred milliseconds, I doubt that it was optimized away.

However, your experiment looks promising. I am just wondering that all the measurements (for example, [that one](https://github.com/hyrise/hyrise/pull/2026#discussion_r400113202)) we discussed in the past did not show any performance improvements, am I wrong?

We have measured traversing the index and since this is a simple binary search on ~65000 values, it was never significant. I believe that we also considered a second TableScan after the IndexScan to see if the use of the SingleChunkPosList is slowing everything down. I don't remember how large the impact was.

Good job! I am curious to see the impact on the TPCHMicroBenchmark.